cloud/pkg/router/listener/http.go

```diff
@@ -0,0 +1,189 @@
+	}
+}
+
+func (rh *RestHandler) AddListener(key interface{}, han Handle) {
+	path, ok := key.(string)
+	if !ok {
+		return
+	}
+
+	rh.handlers.Store(path, han)
+}
+
+func (rh *RestHandler) RemoveListener(key interface{}) {
+	path, ok := key.(string)
+	if !ok {
+		return
+	}
+	rh.handlers.Delete(path)
+}
+
+func (rh *RestHandler) matchedPath(uri string) (string, bool) {
+	var candidateRes string
+	rh.handlers.Range(func(key, value interface{}) bool {
+		pathReg := key.(string)
+		if match := utils.IsMatch(pathReg, uri); match {
+			if candidateRes != "" && utils.RuleContains(strings.Split(pathReg, "/"), strings.Split(candidateRes, "/")) {
+				return true
+			}
+			candidateRes = pathReg
+		}
+		return true
+	})
+	if candidateRes == "" {
+		return "", false
+	}
+	return candidateRes, true
+}
+
+func (rh *RestHandler) httpHandler(w http.ResponseWriter, r *http.Request) {
+	uriSections := strings.Split(r.RequestURI, "/")
+	if len(uriSections) < 2 {
+		//URL format incorrect
+		klog.Warningf("URL format incorrect: %s", r.RequestURI)
+		w.WriteHeader(http.StatusNotFound)
+		_, err := w.Write([]byte("Request error"))
+		if err != nil {
+			klog.Errorf("Response write error: %s, %s", r.RequestURI, err.Error())
+		}
+		return
+	}
+
+	matchPath, exist := rh.matchedPath(r.RequestURI)
+	if !exist {
+		klog.Warningf("URL format incorrect: %s", r.RequestURI)
+		w.WriteHeader(http.StatusNotFound)
+		_, err := w.Write([]byte("Request error"))
+		if err != nil {
+			klog.Errorf("Response write error: %s, %s", r.RequestURI, err.Error())
+		}
+		return
+	}
+	v, ok := rh.handlers.Load(matchPath)
+	if !ok {
+		klog.Warningf("No matched handler for path: %s", matchPath)
+		return
+	}
+	handle, ok := v.(Handle)
+	if !ok {
+		klog.Errorf("invalid conver to Handl. mathch path: %s", matchPath)
+		return
+	}
+	b, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		klog.Errorf("request error, write result: %v", err)
+		w.WriteHeader(http.StatusBadRequest)
+		_, err = w.Write([]byte("Request error,body is null"))
+		klog.Errorf("Response write error: %s, %s", r.RequestURI, err.Error())
+		return
+	}
+
+	if isNodeName(uriSections[1]) {
+		params := make(map[string]interface{})
+		msgID := uuid.New().String()
+		params["messageID"] = msgID
+		params["request"] = r
+		params["timeout"] = rh.restTimeout
+		params["data"] = b
+
+		v, err := handle(params)
+		if err != nil {
+			//w.WriteHeader(http.StatusInternalServerError)
+			//_, err := w.Write([]byte(err.Error()))
+			//klog.Warningf("operation timeout, msg id: %s, write result: %v", msgID, err)
+			return
+		}
+		response, ok := v.(*http.Response)
+		if !ok {
+
+		}
+		body, err := ioutil.ReadAll(response.Body)
+		if err != nil {
+
+		}
+		w.WriteHeader(response.StatusCode)
+		_, err = w.Write(body)
+		if err != nil {
+			klog.Errorf("response body write error, msg id: %s, reason: %s", msgID, err.Error())
+		}
+		klog.Infof("response to client, msg id: %s, write result: success", msgID)
```

Comment:
when err is not nil, this success log is also printed.

Reply:
done